General refactoring and automatic config accumulators

bmhughes opened the pull request:
# Description

- Set resource configuration Hash programmatically
- Automatic config Hash path creation
- Automatic resource property fetching
- Automatic config template resource generation
- Add service resource
- Add LDAP log filters to the main config file

## Issues Resolved

- #380 
- #388
- #392 
- #397 
- #400 
- #401 

## Check List

- [x] All tests pass. See TESTING.md for details.
- [x] New functionality includes testing.
- [x] New functionality has been documented in the README if applicable.

xorima committed:
prep for main

bmhughes committed:
Tidy up templates

bmhughes committed:
Automatic config template resource generation

bmhughes committed:
Test fixup for auto template resources

bmhughes committed:
Resources to unified mode

bmhughes committed:
Fix resource property collision

bmhughes committed:
Add extra_options to grafana_config

Closes #400

bmhughes committed:
Concat to existing settings rather than overwriting them

bmhughes committed:
Add LDAP log filters to the main config file

Fixes #392

bmhughes committed:
Start service immediately during testing

bmhughes committed:
Azure AD auth needs path override

bmhughes committed:
Require Chef 16 for resource partials

bmhughes committed:
Kitchen allow log level to be set

bmhughes committed:
Remove deprecated API resources

bmhughes committed:
Refactor config file accumulation

As we've removed the deprecated API resource then we don't need to worry
about config writing timing so we can use a standard accumulator
pattern to build the config files.

bmhughes committed:
Template nil catch

bmhughes committed:
Delete config_writer resource

bmhughes committed:
Remove redundant sleep test recipe

bmhughes committed:
Remove redundant instance_name

With the migration to an standard accumulator pattern the instance_name
functionality can be achieved by overriding the configuration file or
directory etc.

bmhughes committed:
Library refactoring

bmhughes committed:
Correct config_auth default key path

bmhughes committed:
Test strip global section

bmhughes committed:
Remove etc/share directory creation as we use packages

bmhughes committed:
Resource formatting and update options for Grafana 8.1

jamie-zaharopoulos-fh commented on CHANGELOG.md:
```diff
@@ -4,20 +4,12 @@ This file is used to list changes made in each version of grafana.
-- `grafana_datasource`, `grafana_datasource_backup`
-- `grafana_organization`
-- `grafana_user`
+- Set resource configuration Hash programmatically
```
Was it intentional to remove the changeling for `## 9.7.0 - *2021-07-23*` ?

bmhughes committed:
Unify accumulator config method with action property

bmhughes committed:
Automatic LDAP configuration accumalator template

bmhughes committed:
Install deepsort gem for chefspec

bmhughes committed:
Fixup plug library references

bmhughes committed:
Test refactoring for api resource removal

bmhughes committed:
LDAP config load_current_value support

bmhughes committed:
Split log resource into separate types

bmhughes committed:
Split metrics resource into separate types

bmhughes committed:
Split auth resource into separate types

bmhughes committed:
All resources converge_if_changed_support

bmhughes committed:
Rationalise config resource action naming

bmhughes committed:
Add expression, geomap, feature_toggle and date_format resources

bmhughes committed:
load_current_value for extra_options property

bmhughes committed:
Correct config actions

bmhughes committed:
Remove redundant default values to use grafana defaults instead

bmhughes committed:
Add load_existing option to init_config_template

Allow the loading of the current configuration state to create a
persistent accumulator pattern where configuration is not removed if the
resource creating it is removed.

bmhughes committed:
Move common resource actions into partial

bmhughes committed:
Extract config path override to action_class method

This allows us to use the common resource actions and
load_current_value from the partial in 99% of cases.

bmhughes committed:
Remove header comment

damacus commented on libraries/config.rb:
```diff
@@ -0,0 +1,150 @@
+        Chef::Log.debug("init_config_template: Creating config template resource for #{new_resource.config_file}")
+
+        with_run_context(:root) do
+          declare_resource(:chef_gem, 'deepsort') { compile_time true } unless gem_installed?('deepsort')
```
Do we still need compile time now we're using unified_mode?

lamont-granquist replied:
this is going into the root run_context which doesn't have unified_mode

declaring compile_time resources in the root run context is really weird though, particularly to install gems.  the idempotency check here is also a codesmell.

i don't know why this isn't just `chef_gem 'deepsort'`

if it needs to be done at the outer context with idempotency around the resource and forced to compile time:

`find_resource(:chef_gem, 'deepsort') { compile_time true }`

bmhughes committed:
Load current state of config files into templates on init

bmhughes committed:
Compact and deep sort ini config output

bmhughes committed:
Add debug logging to ldap_config_file helper

bmhughes committed:
Update LDAP resource for load current state changes

bmhughes committed:
Add comments to the libraries

bmhughes committed:
Remove enforce_idempotency overrides from kitchen config

bmhughes committed:
Change skip property constant to method

bmhughes committed:
Global documentation update

bmhughes committed:
Add development guide

bmhughes committed:
Refactor plugin resource to use converge_by

bmhughes committed:
Common config resource delete action

bmhughes committed:
Resource format fixup

bmhughes committed:
Add missing grafanacom path override

bmhughes committed:
Documentation update

bmhughes committed:
Add jaeger and base external image resources

bmhughes committed:
Add missing config resource properties

bmhughes committed:
Add resource property/grafana property translation feature

bmhughes committed:
Markdown link fixup

bmhughes committed:
Add Okta auth resource

bmhughes committed:
Allow overriding of the Grafana package name

Fixes #380

bmhughes committed:
Update README links default branch name

bmhughes committed:
Move LDAP server exist checks into resources

bmhughes committed:
Fix missing prefix delete for ldap attribute resource load_current_value

bmhughes committed:
Add Fedora installation support

damacus commented on DEVELOPING.md:
```diff
@@ -0,0 +1,82 @@
+# Developing
+
+The refactor of the cookbooks in v10 makes some major changes to the underlying method of operation of the cookbook to improve operation and maintainabilty.
```
```suggestion
The refactor for version 10 of this cookbook makes some major changes to the underlying method of operation of the cookbook to improve operation and maintainability.
```

damacus commented on DEVELOPING.md:
```diff
@@ -0,0 +1,82 @@
+
+### Overiding the automatic path
+
+If the name of the resource cannot match the name of the configuration section or a nested configuration section is in use such as `section.subsection` then the automatic configuration path can be overriden by the addition of the `:resource_config_path_override` method. Due to needed to override the path in both the action class and the outer resource class this method must be added to both secti ...
```
```suggestion
If the name of the resource does not match the name of the configuration section or a nested configuration section is in use such as `section.subsection` then the automatic configuration path can be overriden by the use of the `:resource_config_path_override` method. Due to needing to override the path in both the action class and the outer resource class this method must be added to both sections of the resource declaration.
```

damacus commented on DEVELOPING.md:
```diff
@@ -0,0 +1,82 @@
+
+If the name of the resource cannot match the name of the configuration section or a nested configuration section is in use such as `section.subsection` then the automatic configuration path can be overriden by the addition of the `:resource_config_path_override` method. Due to needed to override the path in both the action class and the outer resource class this method must be added to both secti ...
+
+Below the last property add the following with the array contents set to the config Hash path set as you would path to [Hash#dig](https://ruby-doc.org/core-3.0.2/Hash.html#method-i-dig) to create the (nested if required) configuration path.
```
I'm not sure this sentence makes sense right now. Are you able to reword it please?

damacus commented on DEVELOPING.md:
````diff
@@ -0,0 +1,82 @@
+end
+```
+
+### Excluding properties from the configuration
````
This works for this cookbook, but if we're to extract this into a library this doesn't have great UX.

bmhughes replied:
I presume you mean where you have to declare the method twice? I'm just wondering actually if the check for resource vs action class can be done in the helper and the correct method called off that.

damacus replied:
I think what I mean here is having to add the method to the action_class for it to do something isn't super cool. 
I wonder if there's a property we can add. Or maybe I'm just thinking of this from the end user (wrapper cookbook) end which I shouldn't be. 

This property is meant to be for maintainers to add/remove. That said this pattern isn't really in use in any other part of Chef so I think it's caught me off guard slightly!

bmhughes replied:
Yeh it's the latter, you'd only touch this method/hash value when developing the cookbook.

damacus commented on UPGRADING.md:
```diff
@@ -1,5 +1,23 @@
+
+### Version 10 is a major internal rewrite
+
+- All API resource have been removed
```
```suggestion
- All API resources have been removed
```

damacus commented on UPGRADING.md:
```diff
@@ -1,5 +1,23 @@
+### Version 10 is a major internal rewrite
+
+- All API resource have been removed
+- All resource have been refactored around an abstract common base resource `_config_file`
```
```suggestion
- All resources have been refactored around an abstracted common base resource `_config_file`
```

damacus commented on UPGRADING.md:
```diff
@@ -1,5 +1,23 @@
+  - To remove configuration elements an explicit resource with `:delete` action is required
+    - If no properties are specified with the `:delete` action then the whole configuration section will be removed
+    - If properties are specified with the `:delete` action then only the specified properties will be removed
+  - Configuration files will not be overwritten with a blank or partial configuration when a run `raises`
```
```suggestion
  - Configuration files will not be overwritten with a blank or partial configuration when a run fails
```

damacus replied:
I think simply saying fails here is a little easier to understand

damacus commented on UPGRADING.md:
```diff
@@ -1,5 +1,23 @@
+  - Configuration files will not be overwritten with a blank or partial configuration when a run `raises`
+- The auth resource has been split into separate resources
+- The log resource has been split into separate resources
+- The LDAP configuration resource have been split and renamed to singular forms
```
```suggestion
- The LDAP configuration resource has been split and renamed to singular forms
```

bmhughes committed:
Docs fixup

bmhughes committed:
Use tenary for delete properties

bmhughes committed:
Refactor helpers for a single override/translate/skip resource method
